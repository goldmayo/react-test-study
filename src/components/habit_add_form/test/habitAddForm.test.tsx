import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import HabitAddForm from "../habitAddForm";
import TestRenderer from "react-test-renderer";

describe("HabitAddForm", () => {
  it("render", () => {
    const component = TestRenderer.create(<HabitAddForm onAdd={jest.fn()} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
  describe("Form Submit", () => {
    let onAdd: (name: string) => void;
    beforeEach(() => {
      onAdd = jest.fn();
    });
    it("calls onAdd when button is clicked and valid habit is entered", () => {
      render(<HabitAddForm onAdd={onAdd} />);
      const input: HTMLElement = screen.getByPlaceholderText("please enter your habit");
      const button: HTMLElement = screen.getByText("Add");
      //사용자가 input에 원하는 습관의 이름을 입력하고
      userEvent.type(input, "New Habit");
      //add 버튼을 클릭하면
      userEvent.click(button);
      //onAdd가 input에 입력된 이름과 함께 호출되어야 한다
      expect(onAdd).toHaveBeenCalledWith("New Habit");
    });
    it("does not call onAdd when the habit is empty", () => {
      render(<HabitAddForm onAdd={onAdd} />);
      const input: HTMLElement = screen.getByPlaceholderText("please enter your habit");
      const button: HTMLElement = screen.getByText("Add");
      userEvent.click(input);
      userEvent.click(button);
      expect(onAdd).toHaveBeenCalledTimes(0);
    });
  });
});
