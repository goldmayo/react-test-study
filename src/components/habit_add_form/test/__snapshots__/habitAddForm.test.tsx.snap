// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabitAddForm render 1`] = `
<form
  className="add-form"
  onSubmit={[Function]}
>
  <input
    className="add-input"
    placeholder="please enter your habit"
    type="text"
  />
  <button
    className="add-button"
  >
    Add
  </button>
</form>
`;
